sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -719,103 +723,249 @@ static int create_codec_dai_name(struct device *dev,
 		part_id = SDW_PART_ID(adr);
 		class_id = SDW_CLASS_ID(adr);
 
+		comp_index = i + offset;
 		if (is_unique_device(link, sdw_version, mfg_id, part_id,
-				     class_id, i))
-			codec[i].name = devm_kasprintf(dev, GFP_KERNEL,
-						       "sdw:%x:%x:%x:%x",
-						       link_id, mfg_id, part_id,
-						       class_id);
-		else
-			codec[i].name = devm_kasprintf(dev, GFP_KERNEL,
-						       "sdw:%x:%x:%x:%x:%x",
-						       link_id, mfg_id, part_id,
-						       class_id, unique_id);
+				     class_id, i)) {
+			codec_str = "sdw:%x:%x:%x:%x";
+			codec[comp_index].name =
+				devm_kasprintf(dev, GFP_KERNEL, codec_str,
+					       link_id, mfg_id, part_id,
+					       class_id);
+		} else {
+			codec_str = "sdw:%x:%x:%x:%x:%x";
+			codec[comp_index].name =
+				devm_kasprintf(dev, GFP_KERNEL, codec_str,
+					       link_id, mfg_id, part_id,
+					       class_id, unique_id);
+		}
 
-		if (!codec[i].name)
+		if (!codec[comp_index].name)
 			return -ENOMEM;
 
-		id = find_codec_info_part(part_id);
-		if (id < 0)
-			return id;
+		codec_index = find_codec_info_part(part_id);
+		if (codec_index < 0)
+			return codec_index;
 
-		codec[i].dai_name = codec_info_list[id].dai_name;
+		codec[comp_index].dai_name =
+			codec_info_list[codec_index].dai_name;
 	}
 
 	return 0;
 }
 
-static void set_codec_init_func(const struct snd_soc_acpi_link_adr *link,
-				struct snd_soc_dai_link *dai_links,
-				bool playback)
+static int set_codec_init_func(const struct snd_soc_acpi_link_adr *link,
+			       struct snd_soc_dai_link *dai_links,
+			       bool playback)
 {
 	unsigned int part_id;
-	int i, id;
+	int i, codec_index;
 
 	for (i = 0; i < link->num_adr; i++) {
 		part_id = SDW_PART_ID(link->adr_d[i].adr);
-		id = find_codec_info_part(part_id);
+		codec_index = find_codec_info_part(part_id);
 
-		if (codec_info_list[id].init)
-			codec_info_list[id].init(link, dai_links,
-						 &codec_info_list[id],
+		if (codec_index < 0)
+			return codec_index;
+
+		if (codec_info_list[codec_index].init)
+			codec_info_list[codec_index].init(link, dai_links,
+						 &codec_info_list[codec_index],
 						 playback);
 	}
+
+	return 0;
 }
 
-static int create_sdw_codec_dai(struct device *dev, int *be_index,
-				struct snd_soc_dai_link *dai_links,
-				struct snd_soc_dai_link_component *cpus,
-				const struct snd_soc_acpi_link_adr *link,
-				int *cpu_id)
+/*
+ * check endpoint status in slaves and gather link ID for all slaves in
+ * the same group to generate different CPU DAI. Now only support
+ * one sdw link with all slaves set with only single group id.
+ *
+ * one slave on one sdw link without aggregated = 0
+ * one sdw link <---> one-cpu DAI <---> one-codec DAI
+ *
+ * two or more slaves on one sdw link without aggregated = 0:
+ * one sdw link <---> one-cpu DAI <---> multi-codec DAIs
+ *
+ * multiple links with multiple slaves with aggregated = 1:
+ * one sdw link <---> 1 .. N CPU DAIs <----> 1 ,, codec DAIs
```

Comment:
This doesnt make sense? you sai multiple links just above and then you say one sdw link here?
Should this be something like (where N is number of links and M is the number of slaves):
N SDW links <----> 1 .. N CPU DAIs <----> 1 .. M codec dais?

Reply:
yes, thanks!